### ISSUE TITLE
[BUG] ServiceBusSenderClient.sendMessages(Iterable<ServiceBusMessage>) silently drops messages 

### ISSUE THREAD

**Issue by florianmack** [Azure/azure-sdk-for-java#33316]
**Describe the bug**
When using the function com.azure.messaging.servicebus.ServiceBusSenderClient.sendMessages(Iterable<ServiceBusMessage>) we detected that not all of our messages sent to the service bus topic.
Our supposition is that if the size off all message is too big all the other messages are silently dropped

***Exception or Stack Trace***
No Exception is available

**To Reproduce**
Call the function ServiceBusSenderClient.sendMessages(Iterable<ServiceBusMessage>) with a size greater than the maximum size of a single batch 

***Code Snippet***
ServiceBusSenderClient.sendMessages(Iterable<ServiceBusMessage>)

**Expected behavior**
If the size of messages exceed the maximum size of a single batch, an exception will be triggered and the send will fail.

**Screenshots**
Not available

**Setup:**
 - OS: WINDOWS 10
 - IDE: eclipse / intellij
 - Library/Libraries: com.azure:azure-sdk-bom:1.2.9
 - Java version: 11
 - App Server/Environment: Tomcat
 - Frameworks: Spring Boot

**Information Checklist**
- [ x ] Bug Description Added
- [ x ] Repro Steps Added
- [ x ] Setup information Added

**Comment by TheDevOps** [comment]
A quick example to reproduce this observed behavior looks as follows:

Note for all examples: the connection url + topic have to be replaced with something actually correct for them to work
```
ServiceBusSenderClient senderClient = new ServiceBusClientBuilder()
    .connectionString(AzureConfiguration.CONNECTION_URL)
    .sender()
    .topicName(AzureConfiguration.TOPIC)
    .buildClient();
LOGGER.info("Going to send 1000 messages to the topic '{}'", AzureConfiguration.TOPIC);
List<ServiceBusMessage> messages = new ArrayList<>();
for (int j = 1; j <= 1000; j++)
{
    ServiceBusMessage message = new ServiceBusMessage(
        "body46634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz"
            + "544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zg"
            + "tfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghr"
            + "fgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg"
            + "547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547466347"
            + "34ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z5475"
            + "4zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547" + j);
    message.getApplicationProperties().put("longpropertynamethatcausesfurthersizeincrease", j);
    messages.add(message);
}
senderClient.sendMessages(messages);
LOGGER.info("finished sending message batch with size {}", messages.size());
senderClient.close();
```

so it should send 1000 messages in one batch, there is also no error when running it and the final logline is shown as e.g.

```
2023-02-03 09:40:57.759+0100 05 [main] I    at.poc.azure.azure.ProviderClient - finished sending message batch with size 1000
```

however there is another log

```
2023-02-03 09:40:57.592+0100 05 [reactor-executor-1] I  c.a.m.s.ServiceBusSenderAsyncClient - {"az.sdk.message":"Sending batch.","batchSize":355}
```

that looks weird and checking the subscription suddenly shows far less than 1000 messages in it:

![image](https://user-images.githubusercontent.com/44869367/216588805-8f5941de-56f4-4399-9e0e-a01b74a58dec.png)

So only 355 messages, like in the other log, actually arrived. Running it a few times shows it will always be the same number.
So after having a rough guess that it might be related to the size limit another test with only 300 messages sent in the batch to confirm if it's a "static" cutoff or some percentage that is lost as shown below

```
ServiceBusSenderClient senderClient = new ServiceBusClientBuilder()
    .connectionString(AzureConfiguration.CONNECTION_URL)
    .sender()
    .topicName(AzureConfiguration.TOPIC)
    .buildClient();
LOGGER.info("Going to send 1000 messages to the topic '{}'", AzureConfiguration.TOPIC);
List<ServiceBusMessage> messages = new ArrayList<>();
for (int j = 1; j <= 300; j++)
{
    ServiceBusMessage message = new ServiceBusMessage(
        "body46634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz"
            + "544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zg"
            + "tfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghr"
            + "fgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg"
            + "547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547466347"
            + "34ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z5475"
            + "4zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547" + j);
    message.getApplicationProperties().put("longpropertynamethatcausesfurthersizeincrease", j);
    messages.add(message);
}
senderClient.sendMessages(messages);
LOGGER.info("finished sending message batch with size {}", messages.size());
senderClient.close();
```
shows that this time all 300 messages arrived as expected:

![image](https://user-images.githubusercontent.com/44869367/216588959-9c428eb7-12b8-4cf8-b601-bda93635fdc6.png)

To finally further confirm it's likely related to the message size we ran one more test with one additional property set per message to slightly increase the size of each message

```
ServiceBusSenderClient senderClient = new ServiceBusClientBuilder()
    .connectionString(AzureConfiguration.CONNECTION_URL)
    .sender()
    .topicName(AzureConfiguration.TOPIC)
    .buildClient();
LOGGER.info("Going to send 1000 messages to the topic '{}'", AzureConfiguration.TOPIC);
List<ServiceBusMessage> messages = new ArrayList<>();
for (int j = 1; j <= 1000; j++)
{
    ServiceBusMessage message = new ServiceBusMessage(
        "body46634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz"
            + "544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zg"
            + "tfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghr"
            + "fgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg"
            + "547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547466347"
            + "34ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z5475"
            + "4zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547" + j);
    message.getApplicationProperties().put("longpropertynamethatcausesfurthersizeincrease", j);
    messages.add(message);
}
senderClient.sendMessages(messages);
LOGGER.info("finished sending message batch with size {}", messages.size());
senderClient.close();
```

and it shows that this is indeed likely the cause because now an even lower number of messages - 322 - actually arrived

![image](https://user-images.githubusercontent.com/44869367/216588889-359c0449-32ec-4fb5-9063-f0be1328ebe1.png)

So all things considered it seems very likely the client contrary to his documentation of

![image](https://user-images.githubusercontent.com/44869367/216586278-f21ebd46-3b3d-4538-889a-af8253816ac1.png)

does indeed not fail the send but just silently drop messages.

After some checking I've found the following code example https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/src/samples/java/com/azure/messaging/servicebus/SendMessageBatchSample.java that shows the usage of `sender.createMessageBatch(new CreateMessageBatchOptions().setMaximumSizeInBytes(1024))` to address this limit and running a quick test (our max size is 256KB) as shown below

```
ServiceBusSenderClient senderClient = new ServiceBusClientBuilder()
    .connectionString(AzureConfiguration.CONNECTION_URL)
    .sender()
    .topicName(AzureConfiguration.TOPIC)
    .buildClient();
LOGGER.info("Going to send 1000 messages to the topic '{}'", AzureConfiguration.TOPIC);
ServiceBusMessageBatch currentBatch = senderClient.createMessageBatch(
    new CreateMessageBatchOptions().setMaximumSizeInBytes(256*1024));
int count = 0;
for (int j = 1; j <= 1000; j++)
{
    // this examples assumes the topic already exists
    ServiceBusMessage message = new ServiceBusMessage(
        "body46634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz"
            + "544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zg"
            + "tfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghr"
            + "fgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg"
            + "547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547466347"
            + "34ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz54746634734ghrfgdrz544z5475"
            + "4zdhg547zgtfhz54746634734ghrfgdrz544z54754zdhg547zgtfhz547" + j);
    message.getApplicationProperties().put("longpropertynamethatcausesfurthersizeincrease", j);
    if (currentBatch.tryAddMessage(message))
    {
        count++;
        continue;
    }
    // The batch is full, so we create a new batch and send the batch.
    senderClient.sendMessages(currentBatch);
    LOGGER.info("finished sending message batch with size {}", count);
    count = 0;
    currentBatch = senderClient.createMessageBatch();
    // Add that message that we couldn't before.
    if (!currentBatch.tryAddMessage(message)) {
        LOGGER.error("Message is too large for an empty batch. Skipping. Max size: {}. Message: {}",
            currentBatch.getMaxSizeInBytes(), message.getBody());
    }
    else
    {
        count++;
    }

}
senderClient.sendMessages(currentBatch);
LOGGER.info("finished sending message batch with size {}", count);
senderClient.close();
```

which also produces way more favorable logs

```
2023-02-03 10:57:29.753+0100 05 [main] I  c.a.m.s.ServiceBusSenderAsyncClient - {"az.sdk.message":"Sending batch.","batchSize":332}
2023-02-03 10:57:30.033+0100 05 [main] I    at.poc.azure.azure.ProviderClient - finished sending message batch with size 332
2023-02-03 10:57:30.037+0100 05 [main] I  c.a.m.s.ServiceBusSenderAsyncClient - {"az.sdk.message":"Sending batch.","batchSize":331}
2023-02-03 10:57:30.157+0100 05 [main] I    at.poc.azure.azure.ProviderClient - finished sending message batch with size 331
2023-02-03 10:57:30.159+0100 05 [main] I  c.a.m.s.ServiceBusSenderAsyncClient - {"az.sdk.message":"Sending batch.","batchSize":331}
2023-02-03 10:57:30.253+0100 05 [main] I    at.poc.azure.azure.ProviderClient - finished sending message batch with size 331
2023-02-03 10:57:30.254+0100 05 [main] I  c.a.m.s.ServiceBusSenderAsyncClient - {"az.sdk.message":"Sending batch.","batchSize":6}
2023-02-03 10:57:30.323+0100 05 [main] I    at.poc.azure.azure.ProviderClient - finished sending message batch with size 6
```

and finally shows exactly 1000 messages in the subscription.

![image](https://user-images.githubusercontent.com/44869367/216589021-d6ce394b-3b3b-4d91-9c3f-93db26b57ef9.png)

further confirmed our suspicion.

With all this said either the implementation is currently wrong to not raise an error, or the javadoc of the method is not correct because it does not raise one but rather silently drop the messages. 
Personally I'd prefer getting the exception since it's easier to spot and then search and fix with the intended max size batching mechanism rather than just loosing messages without any bad log line ever, but ultimately either one is fine as long as the javadoc is at least correct. Maybe some pointer to `senderClient.createMessageBatch()` in it would also make it more transparent how to use the api in a save way.

Thanks!

**Comment by anuchandy** [comment]
Thanks @florianmack, for the great analysis; I'll see how we can improve the API experience to avoid the undesired dropping of messages.

**Comment by madsvcarlsen** [comment]
Just spent a good while debugging an application in production because of the SDK silently dropping messages.
It should really be prioritized, this issue is coming up on its 1-year birthday, and it makes part of the service bus SDK unusable, and untrustworthy.

We consistently saw sendMessages dropping the same amount of messages when given the same list every time. No log statements were made about dropping the messages, and no exception was thrown.

Introducing our own batch-handling solved the issue.

**Comment by conniey** [comment]
Fixed by #40462
